src/rpc/WorkQueue.hpp

```diff
@@ -19,33 +19,100 @@
 
 #pragma once
 
-#include "util/Assert.hpp"
 #include "util/Mutex.hpp"
-#include "util/Spawn.hpp"
 #include "util/config/ConfigDefinition.hpp"
 #include "util/log/Logger.hpp"
 #include "util/prometheus/Counter.hpp"
 #include "util/prometheus/Gauge.hpp"
 
 #include <boost/asio.hpp>
 #include <boost/asio/spawn.hpp>
+#include <boost/asio/steady_timer.hpp>
+#include <boost/asio/strand.hpp>
 #include <boost/asio/thread_pool.hpp>
 #include <boost/json.hpp>
 #include <boost/json/object.hpp>
 
 #include <atomic>
-#include <chrono>
 #include <cstddef>
 #include <cstdint>
 #include <functional>
 #include <limits>
+#include <queue>
 
 namespace rpc {
 
+namespace impl {
+class OneTimeCallable {
+    std::function<void()> func_;
+    bool called_{false};
+
+public:
+    void
+    setCallable(std::function<void()> func);
+
+    void
+    operator()();
+
+    operator bool() const;
+};
+}  // namespace impl
+
+/**
+ * @brief An interface for any class providing a report as json object
+ */
+struct Reportable {
+    virtual ~Reportable() = default;
+
+    /**
+     * @brief Generate a report of the work queue state.
+     *
+     * @return The report as a JSON object.
+     */
+    [[nodiscard]] virtual boost::json::object
+    report() const = 0;
+};
+
 /**
  * @brief An asynchronous, thread-safe queue for RPC requests.
  */
-class WorkQueue {
+class WorkQueue : public Reportable {
+public:
+    /**
+     * @brief Represents a task scheduling priority
+     */
+    enum class Priority : uint8_t {
+        High,
+        Default,
+    };
+
+private:
+    struct DispatcherState {
+        std::queue<std::function<void(boost::asio::yield_context)>> high;
+        std::queue<std::function<void(boost::asio::yield_context)>> normal;
+
+        bool isIdle = false;
+
+        void
+        push(Priority priority, auto&& task)
+        {
+            if (priority == Priority::High) {
+                high.push(std::forward<decltype(task)>(task));
+            } else {
+                normal.push(std::forward<decltype(task)>(task));
+            }
```

Comment:
nit: you can take a reference to a right queue using a ternary operator, and then push to it